crates/precompile-evm-balances-erc20/src/lib.rs

```diff
@@ -0,0 +1,410 @@
+#[derive(Debug, PartialEq)]
+pub enum Action {
+    /// Returns the name of the token.
+    Name = "name()",
+    /// Returns the symbol of the token.
+    Symbol = "symbol()",
+    /// Returns the decimals places of the token.
+    Decimals = "decimals()",
+    /// Returns the amount of tokens in existence.
+    TotalSupply = "totalSupply()",
+    /// Returns the amount of tokens owned by provided account.
+    BalanceOf = "balanceOf(address)",
+    /// Returns the remaining number of tokens that spender will be allowed to spend on behalf of
+    /// owner through transferFrom. This is zero by default.
+    Allowance = "allowance(address,address)",
+    /// Sets amount as the allowance of spender over the caller’s tokens.
+    Approve = "approve(address,uint256)",
+    /// Moves amount tokens from the caller’s account to recipient.
+    Transfer = "transfer(address,uint256)",
+    /// Moves amount tokens from sender to recipient using the allowance mechanism,
+    /// amount is then deducted from the caller’s allowance.
+    TransferFrom = "transferFrom(address,address,uint256)",
+}
+
+/// Utility alias for easy access to the [`pallet_evm_balances::Config::AccountId`].
+type AccountIdFor<T> = <T as pallet_evm_balances::Config>::AccountId;
+
+/// Utility alias for easy access to the [`pallet_evm_balances::Config::Balance`].
+type BalanceFor<T> = <T as pallet_evm_balances::Config>::Balance;
+
+/// Precompile exposing a `pallet_evm_balance` as an ERC20.
+pub struct EvmBalancesErc20<Runtime, Metadata, GasCost>(PhantomData<(Runtime, Metadata, GasCost)>)
+where
+    GasCost: Get<u64>;
+
+impl<EvmBalancesT, Metadata, GasCost> Precompile
+    for EvmBalancesErc20<EvmBalancesT, Metadata, GasCost>
+where
+    Metadata: Erc20Metadata,
+    EvmBalancesT: pallet_evm_balances::Config,
+    BalanceFor<EvmBalancesT>: Into<U256> + TryFrom<U256>,
+    AccountIdFor<EvmBalancesT>: From<H160>,
+    GasCost: Get<u64>,
+{
+    fn execute(handle: &mut impl PrecompileHandle) -> PrecompileResult {
+        let selector = handle
+            .read_selector()
+            .map_err(|_| PrecompileFailure::Error {
+                exit_status: ExitError::Other("invalid function selector".into()),
+            })?;
+
+        match selector {
+            Action::Name => Self::name(handle),
+            Action::Symbol => Self::symbol(handle),
+            Action::Decimals => Self::decimals(handle),
+            Action::TotalSupply => Self::total_supply(handle),
+            Action::BalanceOf => Self::balance_of(handle),
+            Action::Allowance => Self::allowance(handle),
+            Action::Approve => Self::approve(handle),
+            Action::Transfer => Self::transfer(handle),
+            Action::TransferFrom => Self::transfer_from(handle),
+        }
+    }
+}
+
+impl<EvmBalancesT, Metadata, GasCost> EvmBalancesErc20<EvmBalancesT, Metadata, GasCost>
+where
+    Metadata: Erc20Metadata,
+    EvmBalancesT: pallet_evm_balances::Config,
```

Comment:
Can't we use a generic `Currency` here?

Reply:
Sure, good point, moved to `pallet-token-wrapper` and `Currency` is used as a part of `Config`.